proposals/inline-classes.md

```diff
@@ -0,0 +1,163 @@
+to additional heap allocations.
+
+## Motivation / use cases
+
+Inline classes allow to create wrappers for a value of a certain type and such wrappers would be fully inlined. 
+This is similar to type aliases but inline classes are not assignment-compatible with the corresponding underlying types.
+
+Use cases:
+
+- Unsigned types
+```
+inline class UInt(private val value: Int) { ... }
+inline class UShort(private val value: Short) { ... }
+inline class UByte(private val value: Byte) { ... }
+inline class ULong(private val value: Long) { ... }
+```
+
+- Native types like `size_t` for Kotlin/Native
+- Inline enum classes 
+    - Int enum for [Android IntDef](https://developer.android.com/reference/android/support/annotation/IntDef.html)
+    - String enum for Kotlin/JS (see [WebIDL enums](https://www.w3.org/TR/WebIDL-1/#idl-enums))
+- Units of measurement
+- Result type (aka Try monad) [KT-18608](https://youtrack.jetbrains.com/issue/KT-18608)
+- Inline property delegates
+- Inline wrappers
+
+## Description
+
+Inline classes are declared using soft keyword `inline` and must have a single property:
+```
+inline class Foo(val i: Int)
+```
+Property `i` defines type of the underlying runtime representation for inline class `Foo`, while at compile time type will be `Foo`.
+
+From language point of view, inline classes can be considered as restricted classes, they can declare various members, operators, 
+have generics. 
+
+Example:
+```
+inline class Name(val s: String) : Comparable<Name> {
+    override fun compareTo(other: Name): Int = s.compareTo(other.s)
+    
+    fun greet() {
+        println("Hello, $s")
+    }
+}    
+
+fun greet() {
+    val name = Name("Kotlin") // there is no actual instantiation of class `Name`
+    name.greet() // method `greet` is called as a static method
+}
+```
+
+## Current limitations
+
+Currently, inline classes must satisfy the following requirements:
+
+- Inline class must have a public primary constructor with a single value parameter
+- Inline class must have a single read-only (`val`) property as an underlying value, which is defined in primary constructor
+- Underlying value cannot be of the same type that is containing inline class
+- Inline class cannot have `init` block
+- Inline class must be final
+- Inline class can implement only interfaces
+- Inline class cannot have backing fields
+    - Hence, it follows that inline class can have only simple computable properties (no lateinit/delegated properties)
+- Inline class cannot have inner classes
+- Inline class must be a toplevel class 
+
+### Other restrictions
+
+The following restrictions are related to the usages of inline classes:
+
+- Referential equality (`===`) is prohibited for inline classes
+- vararg of inline class type is prohibited
+```
+inline class Foo(val s: String)
+
+fun test(vararg foos: Foo) { ... } // should be an error  
+```
+
+## Java interoperability  
```

Comment:
I'd also add some short description/examples of how these wrappers look like and where the methods' bodies are